sstables/sstable_set.cc

```diff
@@ -768,6 +768,393 @@ sstable_set_impl::create_single_key_sstable_reader(
+
+    // Used to compare readers stored in _halted_readers.
+    struct halter_reader_cmp {
+        const position_in_partition::less_compare _less;
+
+        explicit halter_reader_cmp(position_in_partition::less_compare cmp)
+            : _less(std::move(cmp)) {}
+
+        bool operator()(reader_iterator a, reader_iterator b) {
+            // Boost heaps are max-heaps, but we want a min-heap, so invert the comparison.
+            return _less(b->sst->max_position(), a->sst->max_position());
+        }
+    };
+
+    const halter_reader_cmp _halted_reader_cmp;
+
+    // In forwarding mode, this is the right-end of the position range being currently queried;
+    // initially it's set to `before_all_clustered_rows` and updated on `fast_forward_to`.
+    // We use it when popping readers from _reader_queue so that we don't prematurely pop
+    // readers that only contain fragments from greater ranges.
+    // In non-forwarding mode _pr_end is always equal to `after_all_clustered_rows`.
+    position_in_partition _pr_end;
+
+    // In forwarding mode, _forwarded_to remembers the last range we were forwarded to.
+    // We need this because we're opening new readers in the middle of the partition query:
+    // after the new reader returns its initial partition-start, we immediately forward it
+    // to this range.
+    std::optional<position_range> _forwarded_to;
+
+    // Remember if we've returned the first fragment, which is always partition-start.
+    // We construct the initial fragment ourselves instead of merging partition-starts returned
+    // from the merged readers; the reason is that we may open new readers *after* returning a partition-start.
+    // We assume that the merged readers don't return partition tombstones, so their partition-starts don't
+    // give any additional information.
+    bool _returned_partition_start = false;
+
+    // In non-forwarding mode, remember if we've returned the last fragment, which is always partition-end.
+    // We construct the fragment ourselves instead of merging partition-ends returned from the merged readers,
+    // because we may close readers in the middle of the partition query.
+    // In forwarding mode this is always false.
+    bool _should_return_partition_end;
+
+    // If a single reader wins with other readers (i.e. returns a smaller fragment) multiple times in a row,
+    // the reader becomes a ``galloping reader'' (and is pointed to by _galloping_reader).
+    // In this galloping mode we stop doing heap operations using the _peeked_readers heap;
+    // instead, we keep peeking the _galloping_reader and compare the returned fragment's position directly
+    // with the fragment of the reader stored at the heap front (if any), hoping that the galloping reader
+    // will keep winning. If he wins, we don't put the fragment on the heap, but immediately return it.
+    // If he loses, we go back to normal operation.
+    reader_iterator _galloping_reader;
```

Comment:
Is this galloping mode expected to be as common as in the case of the other combined reader?

Reply:
I believe it's going to be quite common.
In the previous time windows (not the current one), there should be a single reader, so it's gonna gallop.
Then, when we enter the current time window, we're going to start merging but
1. the sstable overlaps should be small
2. also, the min/max ck metadata is global for the sstable, but we're querying a single partition. So we're often going to have situations where we opened an sstable because it's min/max metadata gives the range [x, y], but for the queried partition it only contains the range [x', y'] where x < x' < y' < y. So in the intervals [x, x') and (y', y] the other readers will be used. I expect there often will be a situation where just a single reader is providing fragments.